doc/barclamp.md

```diff
@@ -0,0 +1,1321 @@
+very straightforward since we only need stub code in there (we normally don't
+do much in Barclamp controllers). This will suffice:
+
+```
+class MybarclampController < BarclampController
+  # Controller for Mybarclamp barclamp
+
+  protected
+
+  def initialize_service
+    @service_object = MybarclampService.new logger
+  end
+end
+```
+
+As you can see, Barbican has the same [piece of boilerplate
+code](https://github.com/crowbar/crowbar-openstack/blob/8bebf8a379ebea8ef462ad49746dda6d36a3c46d/crowbar_framework/app/controllers/barbican_controller.rb).
+
+### Local Testing
+
+*Note: Whenever you make changes during this phase, make them to the
+`crowbar-openstack` git checkout in `/root/crowbar-openstack`, commit them, and
+then run `reposync_crowbar` to make them known to Crowbar.
+
+In the previous sections you created a scaffold for your barclamp, possibly
+with some pieces of implementation as well. Before you create a pull request
+for that code you should test it locally now. While there is a school of
+thought that advocates creating pull requests as soon as the first commit is
+in, this approach is problematic for Crowbar development: one of our CI tests
+is the [Hound](https://houndci.com/) code style checker, which will generate a
+notification for every single violation it detects. Since each commit is likely
+to contain multiple such violations, it is better to get your code working
+first and then present it to Hound at that point. If you test and debug your
+barclamp iteratively while the pull request is already open you will likely
+generate a lot of needless spam (e.g. for code that may end up eventually
+getting removed entirely in the course of development)
+
+For testing your code locally you will need the development environment
+described in the [Prerequisites](#prerequisites) section. So build such an
+environment now (if you haven't already done so) and make sure you've got an
+up-to-date checkout of your `crowbar-openstack` fork in
+`/root/openstack-crowbar` on the development environment's Crowbar admin node.
+This is also a good time to create a LVM snapshot or other backup of your
+Crowbar admin node in case you break it in the course of debugging.
+
+With your Crowbar node backed up and a checkout of your `crowbar-openstack`
+fork in place, apply your code to the running Crowbar instance using the
+commands from the [Preparations for Crowbar Node](#preparations-for-crowbar-node)
+section:
+
+```
+reposync_crowbar   # Fix any merge conflicts reported by this command before proceeding
+```
+
+If you navigate to your Crowbar Admin Node's web interface now and select
+*OpenStack* from the barclamp menu, you should see your new barclamp listed. If
+it doesn't appear or you got an error message from the `sync_crowbar` command
+you will now have to start debugging. Common errors at this stage are mostly
+bad file names, or problems in `mybarclamp.yml`.
+
```

Comment:
I think there's a missing step here, or somewhere around here.  You cover local testing via some mechanism for quickly transferring changes from a git repo into `/opt/dell` (and you suggest `reposync_crowbar` which is one possibility but I have also mentioned others).  This speed of transferral is of course important to keep the hack/test cycle efficient.  So far so good.  But when the changes are stabilising, it then makes sense to sacrifice some development speed for higher quality testing, and in general it's more reliable to test by deploying rpms, which get the test deployment closer to a real world deployment.

The downside of testing via rpm deployments is that it takes a lot longer to build an rpm from a git topic branch and then deploy it to a local test cloud.  But that's exactly why [I automated this process down to a one-line command!](https://github.com/openSUSE/pack-tools/tree/master/contrib/BS-pkg-testing)  So I would like to see that also covered in this guide.